src/gpuarray_reduction.c

```diff
@@ -0,0 +1,548 @@
+	
+	/**
+	 * Outer Loop Header Generation
+	 */
+	
+	for(i=0;i<ctx->numFreeIdx;i++){
+		strb_appendf(s, "\tFOROVER(%d){ESCAPE(%d)\n", i, i);
+	}
+	
+	/**
+	 * Inner Loop Generation
+	 */
+	
+	appendLoopInner(s, ctx);
+	
+	/**
+	 * Outer Loop Trailer Generation
+	 */
+	
+	for(i=0;i<ctx->numFreeIdx;i++){
+		strb_appends(s, "\t}\n");
+	}
+}
+static void  appendLoopInner        (strb*           s,
+                                     GEN_KERNEL_CTX* ctx){
+	int i;
+	
+	/**
+	 * Inner Loop Prologue
+	 */
+	
+	strb_appends(s, "\t/**\n");
+	strb_appends(s, "\t * Reduction initialization.\n");
+	strb_appends(s, "\t */\n");
+	strb_appends(s, "\t\n");
+	
+	appendIdxes (s, "\tT maxV = SRCINDEXER(", "i", 0, ctx->numFreeIdx, "", "");
+	if(ctx->numFreeIdx && ctx->numRedIdx){strb_appends(s, ",");}
+	appendIdxes (s, "", "i", ctx->numFreeIdx, ctx->numIdx, "Start", ");\n");
+	
+	appendIdxes (s, "\tX maxI = RDXINDEXER(", "i", ctx->numFreeIdx, ctx->numIdx, "Start", ");\n");
+	
+	strb_appends(s, "\t\n");
+	strb_appends(s, "\t/**\n");
+	strb_appends(s, "\t * REDUCTION LOOPS.\n");
+	strb_appends(s, "\t */\n");
+	strb_appends(s, "\t\n");
+	
+	/**
+	 * Inner Loop Header Generation
+	 */
+	
+	for(i=ctx->numFreeIdx;i<ctx->numIdx;i++){
+		strb_appendf(s, "\tFOROVER(%d){ESCAPE(%d)\n", i, i);
+	}
+	
+	/**
+	 * Inner Loop Body Generation
+	 */
+	
+	appendIdxes (s, "\tT V = SRCINDEXER(", "i", 0, ctx->numIdx, "", ");\n");
+	strb_appends(s, "\t\n");
+	strb_appends(s, "\tif(V > maxV){\n");
+	strb_appends(s, "\t\tmaxV = V;\n");
+	appendIdxes (s, "\t\tmaxI = RDXINDEXER(", "i", ctx->numFreeIdx, ctx->numIdx, "", ");\n");
+	strb_appends(s, "\t}\n");
+	
+	/**
+	 * Inner Loop Trailer Generation
+	 */
+	
+	for(i=ctx->numFreeIdx;i<ctx->numIdx;i++){
+		strb_appends(s, "\t}\n");
+	}
+	strb_appends(s, "\t\n");
+	
+	/**
+	 * Inner Loop Epilogue Generation
+	 */
+	
+	strb_appends(s, "\t/**\n");
+	strb_appends(s, "\t * Destination writeback.\n");
+	strb_appends(s, "\t */\n");
+	strb_appends(s, "\t\n");
+	appendIdxes (s, "\tDSTMINDEXER(", "i", 0, ctx->numFreeIdx, "", ") = maxV;\n");
+	appendIdxes (s, "\tDSTAINDEXER(", "i", 0, ctx->numFreeIdx, "", ") = maxI;\n");
+}
+static void  appendLoopMacroUndefs  (strb*           s,
+                                     GEN_KERNEL_CTX* ctx){
+	strb_appends(s, "\t#undef FOROVER\n");
+	strb_appends(s, "\t#undef ESCAPE\n");
+	strb_appends(s, "\t#undef SRCINDEXER\n");
+	strb_appends(s, "\t#undef RDXINDEXER\n");
+	strb_appends(s, "\t#undef DSTMINDEXER\n");
+	strb_appends(s, "\t#undef DSTAINDEXER\n");
+}
+
+
+/**
+ * FIXME: Implement a working scheduler and invoker.
+ * 
+ * To schedule effectively the work across several dimensions of possibly
+ * ugly numbers, taking into account the limitations on thread & block
+ * scheduling, will require some library capable of providing a "factoring"
+ * of the tensor dimensions into "nice" prime numbers. Their product may
+ * be allowed to be larger than the original number, provided it remains
+ * within bounds.
+ */
+
```

Comment:
A best-effort approch is OK.  It doesn't have to be perfect on the first try, we can always tune the scheduling later.

Also, I've seen that keeping the number of threads in the 32-256 range is best for CUDA devices.  The number of blocks should ideally be around 16 times the number of MP.  Of course for smaller arrays all these sizes have to be tuned down.